crates/transpiler/src/passes/constrained_reschedule.rs

```diff
@@ -0,0 +1,304 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2025.
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::target::Target;
+use crate::TranspilerError;
+use ::hashbrown::HashSet;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::types::PyDict;
+use pyo3::{pyfunction, wrap_pyfunction, Bound, PyResult};
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Param;
+use qiskit_circuit::operations::{Operation, OperationRef, StandardInstruction};
+use qiskit_circuit::PhysicalQubit;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+use rustworkx_core::petgraph::Direction::Outgoing;
+
+/// Returns the immediate successor operation nodes of a given node in the DAG.
+///
+/// This function traverses the DAG to find all nodes that are direct successors
+/// of the given node and filters them to return only operation nodes.
+///
+/// # Arguments
+///
+/// * `dag` - Reference to the DAGCircuit containing the quantum circuit
+/// * `node_index` - Index of the node whose successors we want to find
+///
+/// # Returns
+///
+/// A vector of `NodeIndex` values representing the immediate successor operation nodes.
+fn get_next_gate(dag: &DAGCircuit, node_index: NodeIndex) -> Vec<NodeIndex> {
+    dag.dag()
+        .neighbors_directed(node_index, Outgoing)
+        .filter(|&succ_idx| {
+            matches!(
+                dag.dag().node_weight(succ_idx),
+                Some(NodeType::Operation(_))
+            )
+        })
+        .collect()
+}
```

Comment:
How does this differ from something like `dag.quantum_successors.filter(|idx| matches!(dag[idx]`, NodeType::Operation(_))`? 

There is also an `op_successors` method on `DAGCircuit` but it's for python. If that is what you're looking for I'd suggest adding it as a `DAGCircuit` method for rust only and work with an iterator return so we can avoid the allocation.